lua/nvim-treesitter-playground/hl-info.lua

```diff
@@ -1,52 +1,83 @@
-local parsers = require "nvim-treesitter.parsers"
-local queries = require'nvim-treesitter.query'
-local ts_utils = require'nvim-treesitter.ts_utils'
-
-local hlmap = vim.treesitter.highlighter.hl_map
+local highlighter = require("vim.treesitter.highlighter")
+local ts_utils = require "nvim-treesitter.ts_utils"
 
 local M = {}
 
-local function is_highlight_name(capture_name)
-  local firstc = string.sub(capture_name, 1, 1)
-  return firstc ~= string.lower(firstc)
-end
-
-function M.show_hl_captures()
-  local bufnr = vim.api.nvim_get_current_buf()
-  local lang = parsers.get_buf_lang(bufnr)
-
-  if not lang then return end
-
+function M.get_treesitter_hl()
+  local buf = vim.api.nvim_get_current_buf()
   local row, col = unpack(vim.api.nvim_win_get_cursor(0))
   row = row - 1
 
-  local parser = parsers.get_parser(bufnr, lang)
-  if not parser then return function() end end
+  local self = highlighter.active[buf]
 
-  parser = parser:language_for_range({row, col, row, col})
+  if not self then return {} end
 
   local matches = {}
-  local query = queries.get_query(parser:lang(), 'highlights')
-
-  for _, tree in ipairs(parser:trees()) do
-    local root = tree:root()
-    local start_row, _, end_row, _ = root:range()
-
-    for _, match in query:iter_matches(root, bufnr, start_row, end_row) do
-      for id, node in pairs(match) do
-        if ts_utils.is_in_node_range(node, row, col) then
-          local c = query.captures[id] -- name of the capture in the query
-          if c ~= nil then
-            table.insert(matches, '@'..c..' -> '..(is_highlight_name(c) and c or (hlmap[c] or 'nil')))
-          end
+
+  self.tree:for_each_tree(function(tstree, tree)
+    if not tstree then return end
+
+    local root = tstree:root()
+    local root_start_row, _, root_end_row, _ = root:range()
+
+    -- Only worry about trees within the line range
+    if root_start_row > row or root_end_row < row then return end
+
+    local query = self:get_query(tree:lang())
+
+    -- Some injected languages may not have highlight queries.
+    if not query:query() then return end
+
+    local iter = query:query():iter_captures(root, self.bufnr, row, row + 1)
+
+    while true do
+      local capture, node = iter()
+      if capture == nil then break end
+
+      local hl = query.hl_cache[capture]
+
+      if hl and ts_utils.is_in_node_range(node, row, col) then
+        local c = query._query.captures[capture] -- name of the capture in the query
+        if c ~= nil then
+          table.insert(matches, "* **@" .. c .. "** -> " .. hl .. " -> **" ..
+                         query:_get_hl_from_capture(capture) .. "**")
```

Comment:
Why show `hl` and `query:_get_hl_from_capture(capture)`? The result should be the same right?

Also the `0` that the query cache uses for no highlight group could need some clarification for the user

Reply:
Ah, it's because of the language specific mappings...

Reply:
Maybe you could still omit one of these if `hl` and `query:_get_hl_from_capture(capture)` are identical (so no actual linking happening)

Reply:
Right, it's indeed different for language specific mappings and thought it was better to always show them, even when different, but I can change it if you want.

Reply:
Fixed